packages/console-types/src/types/provider.ts

```diff
@@ -38,98 +39,128 @@ const providerFunctionalitiesSchema = {
-    urls: {
-      type: 'object',
-      required: ['base', 'apiBase'],
-      properties: {
-        base: { type: 'string' },
-        apiBase: { type: 'string' },
-      },
+    name: { const: GIT_PROVIDER },
+    functionalities: providerFunctionalitiesSchema,
+    repositoryPathTemplate: {
+      type: 'string',
     },
-    base64CA: { type: 'string' },
-    proxy: {
-      type: 'object',
-      properties: {
-        url: { type: 'string' },
-        username: { type: 'string' },
-        password: { type: 'string' },
-      },
-      required: ['url'],
+  },
+} as const
+
+const containerRegistryCapabilitySchema = {
+  type: 'object',
+  additionalProperties: false,
+  required: [
+    'name',
+    'imagePullSecretName',
+    'hostname',
+  ],
+  properties: {
+    name: { const: CONTAINER_REGISTRY },
+    functionalities: providerFunctionalitiesSchema,
+    imagePullSecretName: { type: 'string' },
+    hostname: containerRegistryHostnameString,
+  },
+} as const
+
+const isContainerRegistryCapability = { type: 'object', properties: { name: { const: CAPABILITIES.CONTAINER_REGISTRY } } } as const
+const isGitProviderCapability = { type: 'object', properties: { name: { const: CAPABILITIES.GIT_PROVIDER } } } as const
+
+export const providerCapabilitySchema = {
+  if: isContainerRegistryCapability,
+  then: containerRegistryCapabilitySchema,
+  else: {
+    if: isGitProviderCapability,
+    then: gitProviderCapabilitySchema,
+    else: otherCapabilitySchema,
+  },
+} as const
+
+export const providerCapabilitiesSchema = {
+  type: 'array',
+  items: providerCapabilitySchema,
+} as const
+
+export const providerCommonProperties = {
+  providerId: { type: 'string' },
+  label: { type: 'string' },
+  description: { type: 'string' },
+  type: { type: 'string' },
+  urls: {
+    type: 'object',
+    required: ['base', 'apiBase'],
+    properties: {
+      base: { type: 'string' },
+      apiBase: { type: 'string' },
     },
-    credentials: credentialsSchema,
-    capabilities: {
-      type: 'array',
-      items: providerCapabilitySchema,
+  },
+  base64CA: { type: 'string' },
+  visibility: {
+    additionalProperties: false,
+    type: 'object',
+    properties: {
+      allTenants: { type: 'boolean' },
     },
-    visibility: {
-      additionalProperties: false,
-      type: 'object',
-      properties: {
-        allTenants: { type: 'boolean' },
-      },
+  },
+  proxy: {
+    type: 'object',
+    properties: {
+      url: { type: 'string' },
+      username: { type: 'string' },
+      password: { type: 'string' },
     },
+    required: ['url'],
   },
+  credentials: credentialsSchema,
+  capabilities: providerCapabilitiesSchema,
 } as const
 
-export type ProviderCapability = FromSchema<typeof providerCapabilitySchema>
+export const providerSchema = {
+  type: 'object',
+  additionalProperties: false,
+  properties: providerCommonProperties,
+  required: [
+    'providerId',
+    'type',
+  ],
+} as const
```

Comment:
This will make type narrowing really hard, I'd use a `oneOf` (or `if/then/else`) on the `type` in order to easy the type narrowing.